### PR TITLE
Deprecating Legacy Devices

### PR DESCRIPTION
**Pull request by Shiro-Raven**
Overview of deprecation changes:
- Added general fixture to `conftest.py` for capturing warnings. This should capture any uncaptured warnings from the tests (read, the deprecation warnings from the legacy devices), and therefore works as expected and shouldn't cause any issues.
- Restructured interface tests: moved legacy tests to a separate folder, pulled out tests in the `default.qubit` folder one level, and added missing tests from the legacy tests to the DQ's tests. This should make the removal process much easier by simply deleting a folder and pulling a bunch of files up a directory.
- Updated some examples from the `pulse` module (and related misc. files) that were relying on legacy devices and updated the code block accordingly.
- Switched from `default.qubit.legacy` to `default.mixed` for tests testing the old device interface, and not simply the legacy itself.

[sc-66694[](https://app.shortcut.com/xanaduai/story/66694)]

### CODE REVIEW

File: pennylane/devices/default_qubit_autograd.py [hunk 1]
```diff
@@ -34,6 +37,9 @@ class DefaultQubitAutograd(DefaultQubitLegacy):
 
         pip install autograd
 
+    .. warning::
+        This device is deprecated. Use :class:`~.devices.DefaultQubit` instead; for example through ``qml.device("default.qubit")``.
```
Comment (trbromley) [inline pennylane/devices/default_qubit_autograd.py:41]:
```suggestion
        This device is deprecated. Use :class:`~.devices.DefaultQubit` instead; for example through ``qml.device("default.qubit")``, which now supports backpropagation.
```
For the `default.qubit.torch` etc devices, the main reason to use them is for backprop, so I think it's nice to confirm to the reader here (and in the code warning) that we're not suggesting a regression, just a different way to access.

File: pennylane/devices/default_qubit_autograd.py [hunk 2]
```diff
@@ -104,6 +110,14 @@ def _const_mul(constant, array):
         return constant * array
 
     def __init__(self, wires, *, shots=None, analytic=None):
+        warnings.warn(
+            f"Use of '{self.short_name}' is deprecated. Instead, use 'default.qubit', "
+            "which supports backpropagation. "
+            "If you experience issues, reach out to the PennyLane team on "
+            "the discussion forum: https://discuss.pennylane.ai/",
```
Comment (trbromley) [inline pennylane/devices/default_qubit_autograd.py:117]: Nice, this is a good deprecation to mention the discussion forum.

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -56,6 +56,10 @@
   Instead, use `pennylane.gradients.classical_fisher` and `pennylane.gradients.quantum_fisher`.
   [(#5985)](https://github.com/PennyLaneAI/pennylane/pull/5985)
 
+* The legacy devices `default.qubit.{default.qubit.{legacy,autograd,torch,tf,jax}}` are deprecated.
+  Instead, use ``default.qubit`` as it now supports backpropagation through the several backends.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:60]:
```suggestion
  Instead, use `default.qubit` as it now supports backpropagation through the several backends.
```

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -56,6 +56,10 @@
   Instead, use `pennylane.gradients.classical_fisher` and `pennylane.gradients.quantum_fisher`.
   [(#5985)](https://github.com/PennyLaneAI/pennylane/pull/5985)
 
+* The legacy devices `default.qubit.{default.qubit.{legacy,autograd,torch,tf,jax}}` are deprecated.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:59]:
```suggestion
* The legacy devices `default.qubit.{default.qubit.{autograd,torch,tf,jax,legacy}}` are deprecated.
```
Just to mention `legacy` last because I believe it is the least directly used.

File: pennylane/devices/default_qubit_autograd.py [hunk 5]
```diff
@@ -34,6 +37,9 @@ class DefaultQubitAutograd(DefaultQubitLegacy):
 
         pip install autograd
 
+    .. warning::
+        This device is deprecated. Use :class:`~.devices.DefaultQubit` instead; for example through ``qml.device("default.qubit")``, which now supports backpropagation.
```
Comment (mudit2812) [inline pennylane/devices/default_qubit_autograd.py:41]:
```suggestion
        This device is deprecated. Use :class:`~pennylane.devices.DefaultQubit` instead; for example through ``qml.device("default.qubit")``, which now supports backpropagation.
```

Can you make the same change for the other devices too?

File: pennylane/devices/default_qubit_legacy.py [hunk 6]
```diff
@@ -78,13 +79,13 @@ def _get_slice(index, axis, num_axes):
...
+    r"""Default qubit device for PennyLane.
 
     .. warning::
 
-        This is the legacy implementation of DefaultQubit. It has been replaced by
-        ``qml.devices.DefaultQubit``, which can be accessed with the familiar constructor,
-        ``qml.device("default.qubit")``.
+        This is the legacy implementation of DefaultQubit and is deprecated. It has been replaced by
+        :class:`~.devices.DefaultQubit`, which can be accessed with the familiar constructor,
+        ``qml.device("default.qubit")``, and now supports backpropagation.
```
Comment (mudit2812) [inline pennylane/devices/default_qubit_legacy.py:88]: Why is backprop support in `DefaultQubit` highlighted so consistently? It's been around forever so it seems pointless.

File: pennylane/pulse/hardware_hamiltonian.py [hunk 7]
```diff
@@ -104,9 +106,9 @@ def circuit(params):
 
     >>> params = [2.4]
     >>> circuit(params)
-    Array(0.32495208, dtype=float64)
+    Array(-0.17375104, dtype=float64)
```
Comment (mudit2812) [inline pennylane/pulse/hardware_hamiltonian.py:109]: Why did these values change? We're not executing with shots so there shouldn't be any stochasticity.

File: pennylane/pulse/parametrized_evolution.py [hunk 8]
```diff
@@ -320,11 +322,11 @@ def circuit(param, time):
...
-               [0.9897738 , 0.01022595],
-               [0.9599043 , 0.04009585],
-               [0.9123617 , 0.08763832],
-               [0.84996957, 0.15003097],
-               [0.7761489 , 0.22385144]], dtype=float32)
+            [0.98977406, 0.01022594],
+            [0.95990416, 0.04009584],
+            [0.91236167, 0.08763833],
+            [0.84996865, 0.15003133],
+            [0.77614817, 0.22385181]], dtype=float64)
```
Comment (mudit2812) [inline pennylane/pulse/parametrized_evolution.py:329]: Why was this formatting changed? Running this locally, the old formatting seems consistent with what is being displayed.

File: tests/drawer/test_tape_mpl.py [hunk 9]
```diff
@@ -974,11 +983,11 @@ def test_single_mcm_measure(self):
 
         # probably not a good way to test this, but the best I can figure out
         assert pe1._gc == {
-            "linewidth": 5 * plt.rcParams["lines.linewidth"],
+            "linewidth": 5 * 1.5,  # lines.linewidth for black white style
             "foreground": "black",  # lines.color for black white style
         }
         assert pe2._gc == {
-            "linewidth": 3 * plt.rcParams["lines.linewidth"],
+            "linewidth": 3 * 1.5,  # lines.linewidth for black white style
```
Comment (mudit2812) [inline tests/drawer/test_tape_mpl.py:990]: Why was this changed? Seems like better testing practice to me to not use hard-coded values.

File: tests/test_qnode.py [hunk 10]
```diff
@@ -896,11 +893,12 @@ def circuit(x, y):
         ],
     )
     def test_defer_meas_if_mcm_unsupported(
-        self, dev, first_par, sec_par, return_type, mv_return, mv_res, mocker
+        self, dev_name, first_par, sec_par, return_type, mv_return, mv_res, mocker
     ):  # pylint: disable=too-many-arguments
         """Tests that the transform using the deferred measurement principle is
         applied if the device doesn't support mid-circuit measurements
         natively."""
+        dev = qml.device(dev_name, wires=3)
```
Comment (mudit2812) [inline tests/test_qnode.py:901]: Shouldn't the warning be getting filtered here?

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Looks good. I'm just concerned about the drawer tests. If the deprecation caused changes implicitly in the drawer, I think it makes sense to fix those before merging this.

File: tests/test_qnode.py (hunk 10 above)
Comment (Shiro-Raven) [reply tests/test_qnode.py:901]: No cuz it would be captured by the fixture

File: pennylane/pulse/hardware_hamiltonian.py (hunk 7 above)
Comment (Shiro-Raven) [reply pennylane/pulse/hardware_hamiltonian.py:109]: I re-ran the old example and it seemed to be giving the same values, so I am not sure where the original values are coming from in the first place

File: tests/drawer/test_tape_mpl.py (hunk 9 above)
Comment (Shiro-Raven) [reply tests/drawer/test_tape_mpl.py]: My latest push fixed those issues. I reverted the changes 👌

File: pennylane/devices/default_qubit_legacy.py [hunk 11]
```diff
@@ -78,13 +79,13 @@ def _get_slice(index, axis, num_axes):
...
-    """Default qubit device for PennyLane.
+    r"""Default qubit device for PennyLane.
 
     .. warning::
 
-        This is the legacy implementation of DefaultQubit. It has been replaced by
-        ``qml.devices.DefaultQubit``, which can be accessed with the familiar constructor,
-        ``qml.device("default.qubit")``.
+        This is the legacy implementation of DefaultQubit and is deprecated. It has been replaced by
+        :class:`~.devices.DefaultQubit`, which can be accessed with the familiar constructor,
```
Comment (mudit2812) [inline pennylane/devices/default_qubit_legacy.py:87]:
```suggestion
        :class:`~pennylane.devices.DefaultQubit`, which can be accessed with the familiar constructor,
```

**Review by mudit2812 — APPROVED** [review]
Looks good. Minor non-blocking comment above.

File: tests/interfaces/default_qubit_integration/test_tensorflow.py
Comment (albi3ro) [inline tests/interfaces/default_qubit_integration/test_tensorflow.py:1]: Sorry about continually putting off renaming this folder for too long. 

These tests are no longer specific to just `default.qubit`, and now also test lightning qubit and other devices, I'd support moving these tests to simply `tests/interfaces/test_*.py`.  Basically, we just invert the nesting for the new and the legacy device test folders.

Since we are already moving these files, I might also propose moving them to `tests/workflow/interfaces`.

File: tests/test_debugging.py [hunk 12]
```diff
@@ -273,48 +273,6 @@ def circuit():
 
         _compare_numpy_dicts(result, expected)
 
-    @pytest.mark.parametrize("diff_method", [None, "backprop", "parameter-shift", "adjoint"])
```
Comment (albi3ro) [inline tests/test_debugging.py:276]: I'd say we should probably keep this test around and only remove it when we remove `default.qubit.legacy`.

**Review by albi3ro — APPROVED** [review]
This makes me so happy :)
